src/libopensc/pkcs15-pubkey.c

```diff
@@ -1499,76 +1468,127 @@ static struct ec_curve_info {
 		{"secp521r1",		"1.3.132.0.35", "06052B81040023", 521},
 		{"nistp521",		"1.3.132.0.35", "06052B81040023", 521},
 
-		{"brainpoolP192r1",	"1.3.36.3.3.2.8.1.1.3", "06092B2403030208010103", 192},
-		{"brainpoolP224r1",	"1.3.36.3.3.2.8.1.1.5", "06092B2403030208010105", 224},
-		{"brainpoolP256r1",	"1.3.36.3.3.2.8.1.1.7", "06092B2403030208010107", 256},
-		{"brainpoolP320r1",	"1.3.36.3.3.2.8.1.1.9", "06092B2403030208010109", 320},
+		{"brainpoolP192r1",	"1.3.36.3.3.2.8.1.1.3",  "06092B2403030208010103", 192},
+		{"brainpoolP224r1",	"1.3.36.3.3.2.8.1.1.5",  "06092B2403030208010105", 224},
+		{"brainpoolP256r1",	"1.3.36.3.3.2.8.1.1.7",  "06092B2403030208010107", 256},
+		{"brainpoolP320r1",	"1.3.36.3.3.2.8.1.1.9",  "06092B2403030208010109", 320},
 		{"brainpoolP384r1",	"1.3.36.3.3.2.8.1.1.11", "06092B240303020801010B", 384},
 		{"brainpoolP512r1",	"1.3.36.3.3.2.8.1.1.13", "06092B240303020801010D", 512},
 
 		{"secp192k1",		"1.3.132.0.31", "06052B8104001F", 192},
 		{"secp256k1",		"1.3.132.0.10", "06052B8104000A", 256},
 
-		{"ed25519",		"1.3.6.1.4.1.11591.15.1", "06092B06010401DA470F01", 255},
+		/* OpenPGP extensions by Yubikey and GNUK are not defined in RFCs but we know the oid written to card */
+
+		{"edwards25519",	"1.3.6.1.4.1.11591.15.1", "06092B06010401DA470F01",   255},
 		{"curve25519",		"1.3.6.1.4.1.3029.1.5.1", "060A2B060104019755010501", 255},
 
+		/* RFC 8410 defined curves */
+		{"X25519",              "1.3.101.110", "06032b656e", 255},
+		{"X448",		"1.3.101.111", "06032b656f", 448},
+		{"Ed25519",             "1.3.101.112", "06032b6570", 255},
+		{"Ed448",		"1.3.101.113", "06032b6571", 448},
+		/* GnuPG openpgp curves as used in gnupg-card are equivalent to RFC8410 OIDs */
+		{"cv25519",		"1.3.101.110", "06032b656e", 255},
+		{"ed25519",		"1.3.101.112", "06032b6570", 255},
+
 		{NULL, NULL, NULL, 0}, /* Do not touch this */
 };
-
+// clang-format on
 
 int
 sc_pkcs15_fix_ec_parameters(struct sc_context *ctx, struct sc_ec_parameters *ecparams)
 {
 	int rv, ii;
+	int mapped = 0;
 
 	LOG_FUNC_CALLED(ctx);
 
 	/* In PKCS#11 EC parameters arrives in DER encoded form */
-	if (ecparams->der.value && ecparams->der.len)   {
-		for (ii=0; ec_curve_infos[ii].name; ii++)   {
-			struct sc_object_id id;
-			unsigned char *buf = NULL;
-			size_t len = 0;
-
-			sc_format_oid(&id, ec_curve_infos[ii].oid_str);
-			sc_encode_oid (ctx, &id, &buf, &len);
+	if (ecparams->der.value && ecparams->der.len && ecparams->der.len > 2) {
+		/* caller provided a der version of OID */
+		switch (ecparams->der.value[0]) {
```

Comment:
Instead of handling the ASN.1 buffer by hand in `case 0x13`, please use sc_asn1_read_tag directly.

Reply:
The code has been rewritten, so this does not apply.